src/tpm2_param_enc.c

```diff
@@ -109,82 +107,77 @@ static int TPM2_KDFa(
     if (hLen <= 0)
         return NOT_COMPILED_IN;
 
-    ret = wc_HmacInit(&hmac, NULL, INVALID_DEVID);
-    if (ret != 0)
-        return ret;
+    /* determine total output length */
+    outLen = ((sizeInBits + 7) / 8);
 
-    /* setup counter */
-    counter = (counterInc != NULL) ? *counterInc : 0;
+    /* setup counter - KDFa counter starts at 1 */
+    counter = (counterInc != NULL) ? *counterInc : 1;
 
-    /* get label length if provided */
-    if (label != NULL)
-        lLen = (int)XSTRLEN(label);
-
-    /* generate required bytes */
-    outLen = doOnce ? hLen : (((int)sizeInBits + 7) / 8);
-    outStream = keyStream;
+    /* get label length if provided, including null termination */
+    if (label != NULL) {
+        lLen = (int)XSTRLEN(label) + 1;
+    }
 
     /* Do we have unbounded, unsalted session? */
     if (key == NULL) {
         /* Then, use the authValue to gen session key */
         key = (TPM2B_DATA *)authValue;
     }
 
-    for (; outLen > 0; outLen -= hLen) {
-        if (hLen > outLen)
-            hLen = outLen;
-
-        counter++;
+    ret = wc_HmacInit(&hmac, NULL, INVALID_DEVID);
+    if (ret != 0)
+        return ret;
 
+    /* generate required bytes */
+    for (pos = 0; pos < outLen; pos += hLen) {
         /* start HMAC */
         ret = wc_HmacSetKey(&hmac, hashType, &key->buffer[0], key->size);
         if (ret != 0)
-            break;
+            goto exit;
 
-        /* add counter */
+        /* add counter - KDFa i2 */
         TPM2_Packet_U32ToByteArray(counter, uint32Buf);
         ret = wc_HmacUpdate(&hmac, uint32Buf, (word32)sizeof(uint32Buf));
         if (ret != 0)
-            break;
+            goto exit;
 
-        /* add label */
+        /* add label - KDFa label */
         if (label != NULL) {
             ret = wc_HmacUpdate(&hmac, (byte*)label, lLen);
             if (ret != 0)
-                break;
+                goto exit;
         }
 
         /* add contextU */
         if (contextU != NULL) {
             ret = wc_HmacUpdate(&hmac, contextU->buffer, contextU->size);
             if (ret != 0)
-                break;
+                goto exit;
         }
 
         /* add contextV */
         if (contextV != NULL) {
             ret = wc_HmacUpdate(&hmac, contextV->buffer, contextV->size);
             if (ret != 0)
-                break;
+                goto exit;
         }
 
 
         /* add size in bits */
         TPM2_Packet_U32ToByteArray(sizeInBits, (byte*)&uint32Buf);
         ret = wc_HmacUpdate(&hmac, uint32Buf, (word32)sizeof(uint32Buf));
         if (ret != 0)
-            break;
+            goto exit;
 
         /* get result */
-        ret = wc_HmacFinal(&hmac, outStream);
+        ret = wc_HmacFinal(&hmac, keyStream);
         if (ret != 0)
-            break;
+            goto exit;
 
-        outStream = &outStream[hLen];
+        counter++;
+        keyStream += hLen;
```

Comment:
Something missing here for handling the remainder portion that is not digest sized. Let’s discuss Tue.